### PR TITLE
xcute: Handle backend errors

### PR DESCRIPTION
**Pull request by AymericDu**
##### SUMMARY

Handle backend errors

##### ISSUE TYPE

 - Bugfix Pull Request

##### COMPONENT NAME

- `xcute`

##### SDS VERSION

```
openio 5.2.4.dev114
```

### CODE REVIEW

File: oio/xcute/orchestrator.py [hunk 1]
```diff
@@ -80,22 +93,23 @@ def run_forever(self):
...
-
         # start processing replies
         listen_thread = threading.Thread(target=self.listen)
         listen_thread.start()
-
         self.threads[listen_thread.ident] = listen_thread
 
+        # restart running jobs
+        orchestrator_jobs = None
+        if self.running:
```
Comment (loamhoof) [inline oio/xcute/orchestrator.py]: Early return
There is no point putting everything inside ifs

File: oio/xcute/orchestrator.py [hunk 2]
```diff
@@ -59,6 +61,17 @@ def __init__(self, conf, logger=None):
...
 
+    def handle_backend_errors(self, func, *args, **kwargs):
+        while True:
+            try:
+                return func(*args, **kwargs)
+            except (RedisConnectionError, RedisTimeoutError) as exc:
+                self.logger.error(
+                    'Fail to communicate with redis: %s', exc)
+                if not self.running:
+                    return
```
Comment (loamhoof) [inline oio/xcute/orchestrator.py]: You're changing the return value of `func`
This is risky

File: oio/xcute/orchestrator.py [hunk 3]
```diff
@@ -203,9 +218,11 @@ def dispatch_job(self, job_id, job_type, job_config, job_tasks,
                     beanstalkd_workers,
                     job_id, job_type, job_config, tasks)
                 if sent:
-                    job_status = self.backend.update_tasks_sent(
-                        job_id, tasks.keys())
+                    job_status = self.handle_backend_errors(
+                        self.backend.update_tasks_sent, job_id, tasks.keys())
                     tasks.clear()
+                    if job_status is None:  # self.running == False
```
Comment (loamhoof) [inline oio/xcute/orchestrator.py]: Consequence of changing the return value of the function.

File: oio/xcute/orchestrator.py [hunk 4]
```diff
@@ -126,7 +141,7 @@ def orchestrate_loop(self):
             except Exception:
                 self.logger.exception(
                     'Failed to instantiate job %s', job_id)
-                self.backend.fail(job_id)
+                self.handle_backend_errors(self.backend.fail(job_id))
```
Comment (loamhoof) [inline oio/xcute/orchestrator.py]: That doesn't work.

File: oio/xcute/orchestrator.py [hunk 5]
```diff
@@ -117,7 +131,8 @@ def orchestrate_loop(self):
         """
 
         new_jobs = iter(
-            lambda: self.backend.run_next(self.orchestrator_id), None)
+            lambda: self.handle_backend_errors(
+                self.backend.run_next(self.orchestrator_id), None))
```
Comment (loamhoof) [inline oio/xcute/orchestrator.py]: That doesn't work.

File: oio/xcute/orchestrator.py [hunk 6]
```diff
@@ -216,21 +233,22 @@ def dispatch_job(self, job_id, job_type, job_config, job_tasks,
...
-                    return
-
-            self.backend.free(job_id)
-        except Exception:
-            self.logger.exception('Failed generating task list (job_id=%s)',
-                                  job_id)
+                    job_status = self.handle_backend_errors(
+                        self.backend.update_tasks_sent, job_id, tasks.keys(),
+                        all_tasks_sent=True)
+                    if job_status is not None:  # self.running == True
```
Comment (loamhoof) [inline oio/xcute/orchestrator.py]: Same

**Comment by loamhoof** [comment]
Overall a pretty heavy change for the code. Can't we just catch the exception at a higher level and put the retry at a lower one?

File: oio/xcute/orchestrator.py (hunk 1 above)
Comment (AymericDu) [reply oio/xcute/orchestrator.py]: If we return here, we don't wait for `beanstalkd_thread` and `listen_thread` to stop. To exit properly, we have to wait for all started threads to stop.

Comment (loamhoof) [reply oio/xcute/orchestrator.py]: @AymericDu You can put this code inside a function and wait for the threads to stop outside the function?

File: oio/xcute/orchestrator.py [hunk 7]
```diff
@@ -116,8 +136,15 @@ def orchestrate_loop(self):
...
-            lambda: self.backend.run_next(self.orchestrator_id), None)
+        def run_next_job():
+            next_job, exc = self.handle_backend_errors(
+                self.backend.run_next, self.orchestrator_id)
+            if exc is None:
+                return next_job
+            self.logger.warn('Unable to run next job: %s', exc)
+            return None
+
+        new_jobs = iter(lambda: run_next_job(), None)
```
Comment (loamhoof) [inline oio/xcute/orchestrator.py]: If the code has become this complicated, I don't think there's any point in using `iter(iterable, None)` anymore, is it?

**Pull request merged**